optuna/storages/in_memory.py

```diff
@@ -205,165 +230,213 @@ def set_trial_state(self, trial_id, state):
+
         with self._lock:
-            self.check_trial_is_updatable(trial_id, self.trials[trial_id].state)
+            self.check_trial_is_updatable(trial_id, trial.state)
 
-            trial = copy.copy(self.trials[trial_id])
-            values = copy.copy(self.trials[trial_id].intermediate_values)
+            trial = copy.copy(trial)
+            values = copy.copy(trial.intermediate_values)
             if step in values:
                 return False
 
             values[step] = intermediate_value
             trial.intermediate_values = values
-            self.trials[trial_id] = trial
+            self._set_trial(trial_id, trial)
 
             return True
 
     def set_trial_user_attr(self, trial_id, key, value):
         # type: (int, str, Any) -> None
 
+        self._check_trial_id(trial_id)
+        trial = self.get_trial(trial_id)
+        self.check_trial_is_updatable(trial_id, trial.state)
+
         with self._lock:
-            self.check_trial_is_updatable(trial_id, self.trials[trial_id].state)
+            self.check_trial_is_updatable(trial_id, trial.state)
 
-            trial = copy.copy(self.trials[trial_id])
+            trial = copy.copy(trial)
             trial.user_attrs = copy.copy(trial.user_attrs)
             trial.user_attrs[key] = value
-            self.trials[trial_id] = trial
+            self._set_trial(trial_id, trial)
 
     def set_trial_system_attr(self, trial_id, key, value):
         # type: (int, str, Any) -> None
 
+        trial = self.get_trial(trial_id)
+        self.check_trial_is_updatable(trial_id, trial.state)
+
         with self._lock:
-            self.check_trial_is_updatable(trial_id, self.trials[trial_id].state)
+            self.check_trial_is_updatable(trial_id, trial.state)
 
-            trial = copy.copy(self.trials[trial_id])
+            trial = copy.copy(trial)
             trial.system_attrs = copy.copy(trial.system_attrs)
             trial.system_attrs[key] = value
-            self.trials[trial_id] = trial
+            self._set_trial(trial_id, trial)
 
     def get_trial(self, trial_id):
         # type: (int) -> FrozenTrial
 
-        with self._lock:
-            return copy.deepcopy(self.trials[trial_id])
+        self._check_trial_id(trial_id)
+        study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
+        return self._studies[study_id].trials[trial_number]
+
+    def _set_trial(self, trial_id: int, trial: FrozenTrial) -> None:
+        study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
+        self._studies[study_id].trials[trial_number] = trial
 
     def get_all_trials(self, study_id, deepcopy=True):
         # type: (int, bool) -> List[FrozenTrial]
 
         self._check_study_id(study_id)
         with self._lock:
             if deepcopy:
-                return copy.deepcopy(self.trials)
+                return [copy.deepcopy(trial) for trial in self._studies[study_id].trials]
```

Comment:
I'm just curious: are there any reasons you prefer the current code to `copy.deepcopy(self._studies[study_id].trials)`?

Reply:
Nothing. I remember that I intentionally used the verbose style because the original code did it, but it seems it's just my misunderstanding.